services/api/blocksim_ratelimiter.go

```diff
@@ -101,9 +107,19 @@ func SendJSONRPCRequest(client *http.Client, req jsonrpc.JSONRPCRequest, url str
 	}
 	defer resp.Body.Close()
 
+	// read all resp bytes
+	rawResp, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read response bytes: %w", err), nil
+	}
+
+	// try json parsing
 	res = new(jsonrpc.JSONRPCResponse)
-	if err := json.NewDecoder(resp.Body).Decode(res); err != nil {
-		return nil, err, nil
+	if err := json.NewDecoder(bytes.NewReader(rawResp)).Decode(res); err != nil {
+		// JSON parsing didn't work, return *jsonrpc.JSONRPCResponse with full response for debugging
+		res.Error = &jsonrpc.JSONRPCError{ //nolint:exhaustruct
+			Message: fmt.Errorf("json error: %v", string(rawResp[:])).Error(), //nolint:goerr113
```

Comment:
Is this not equivalent to this?

```suggestion
			Message: fmt.Sprintf("json error: %v", string(rawResp[:]))
```

Reply:
good point! done :-)